### PR TITLE
added tracing capability to ext-authz http client

### LINKED ISSUE #6520: B3 HTTP headers not propagated to ext_authz server in 1.10.0
*B3 HTTP headers are not propagated to ext_authz server*

*Description*:
Upgrading from Envoy 1.9.0 to 1.10.0 omits the Zipkin trace context propagation (B3 headers) to an external HTTP authorization server and hence is not visible in a distributed trace.

Using Envoy 1.9.0 the following config propagates B3 trace context (x-b3-traceid and x-b3-spanid headers) to an external auth server so it can join a distributed trace.
```
- name: envoy.ext_authz
     config:
       http_service:
         server_uri:
           uri: http://authorization-service:8080
           cluster: ext-authz
           timeout: 2s
         allowed_request_headers: ["x-b3-traceid", "x-b3-spanid"]
```

With Envoy 1.10.0 this configuration has changed (see https://www.envoyproxy.io/docs/envoy/v1.10.0/intro/version_history) and the corresponding config looks like
```
- name: envoy.ext_authz
     config:
       http_service:
         server_uri:
           uri: http://authorization-service:8080
           cluster: ext-authz
           timeout: 2s
         authorization_request:
           allowed_headers:
             patterns:
               prefix: "x-b3"
```
However, in 1.10.0 none of the x-b3-* keys are added in HTTP headers and hence the external authz. server can't be part of the trace. I've tried several patterns (exact, prefix, regex) to see if there are any issues with the StringMatcher but without any luck :(

Note that it is only the external auth.server that doesn't get the HTTP headers, I can see the trace which is initiated by Envoy in my tracing UI (Jeager) but without the external auth.server so some parts of the trace is working as expected.

### PR DESCRIPTION
**Pull request by gsagula**
Description:
This PR adds tracing support to the HTTP client in the external authorization filter. So far, only gRPC client was able to trace requests from the filter to an authorization service.

Risk Level: Low
Testing: yes
Docs Changes: N/A
Release Notes: Added
Fixes #6520

Signed-off-by: Gabriel Linden Sagula gsagula@gmail.com

### CODE REVIEW

**Comment by gsagula** [comment]
@dio Sorry for the delay on that. I think I addressed all your comments from https://github.com/envoyproxy/envoy/pull/7267. Thanks!

**Comment by dio** [comment]
Thanks. will take a look.

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc [hunk 1]
```diff
@@ -150,23 +152,30 @@ Http::LowerCaseStrPairVector ClientConfig::toHeadersAdd(
...
-    : cm_(cm), config_(config) {}
+RawHttpClientImpl::RawHttpClientImpl(Upstream::ClusterManager& cm, ClientConfigSharedPtr config,
+                                     TimeSource& time_source)
+    : cm_(cm), config_(config), time_source_(time_source) {}
 
 RawHttpClientImpl::~RawHttpClientImpl() { ASSERT(!callbacks_); }
 
 void RawHttpClientImpl::cancel() {
   ASSERT(callbacks_ != nullptr);
+  span_->setTag(Tracing::Tags::get().Status, Tracing::Tags::get().Canceled);
```
Comment (zuercher) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: nit: `ASSERT(span_ != nullptr)`

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc [hunk 2]
```diff
@@ -150,23 +152,30 @@ Http::LowerCaseStrPairVector ClientConfig::toHeadersAdd(
...
+  span_->finishSpan();
   request_->cancel();
   callbacks_ = nullptr;
```
Comment (zuercher) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: Should we `span_ = nullptr` as well and assert that it's null in the destructor?

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc [hunk 3]
```diff
@@ -209,12 +218,15 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
 
 void RawHttpClientImpl::onSuccess(Http::MessagePtr&& message) {
   callbacks_->onComplete(toResponse(std::move(message)));
+  span_->finishSpan();
```
Comment (zuercher) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: Add `span_ = nullptr`?

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc [hunk 4]
```diff
@@ -209,12 +218,15 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
...
   callbacks_ = nullptr;
 }
 
 void RawHttpClientImpl::onFailure(Http::AsyncClient::FailureReason reason) {
   ASSERT(reason == Http::AsyncClient::FailureReason::Reset);
   callbacks_->onComplete(std::make_unique<Response>(errorResponse()));
+  span_->setTag(Tracing::Tags::get().Error, Tracing::Tags::get().True);
+  span_->finishSpan();
```
Comment (zuercher) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc]: Add `span_ = nullptr`?

**Comment by gsagula** [comment]
@zuercher @dio Thank you for reviewing it.

**Comment by mattklein123** [comment]
Needs master merge.

**Comment by gsagula** [comment]
Done.

**Comment by mattklein123** [comment]
@zuercher I think this just needs a re-approve but I haven't been tracking carefully.

**Pull request merged**